### PR TITLE
add undefined behavior sanitizer

### PR DESCRIPTION
**Pull request by Mic92**
in addition to address sanitizer the undefined behavior sanitizer also catches bugs like integer overflows.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -1,7 +1,13 @@
 AM_CXXFLAGS = -Wall -std=c++11 -D_FILE_OFFSET_BITS=64
 
+SAN_FLAGS = -fno-optimize-sibling-calls -fsanitize-address-use-after-scope -fno-omit-frame-pointer -g -O1
```
Comment (Mic92) [inline src/Makefile.am:3]: @andir thoughts?

File: src/Makefile.am [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
 if WITH_ASAN
-AM_CXXFLAGS += -fsanitize=address -fno-optimize-sibling-calls -fsanitize-address-use-after-scope -fno-omit-frame-pointer -g -O1
+AM_CXXFLAGS += -fsanitize=address
+endif
+
+if WITH_UBSAN
+AM_CXXFLAGS += -fsanitize=undefined
+endif
+
+SAN_FLAGS = -fno-optimize-sibling-calls -fsanitize-address-use-after-scope -fno-omit-frame-pointer
```
Comment (Mic92) [inline src/Makefile.am:11]: Turns out that `-g -O2` is already part of the debug flags, so adding it here has no effect.

File: flake.nix [hunk 3]
```diff
@@ -58,13 +58,18 @@
           };
 
         build = forAllSystems (system: nixpkgsFor.${system}.patchelf-new);
+        build-sanitized = forAllSystems (system: nixpkgsFor.${system}.patchelf-new.overrideAttrs (old: {
```
Comment (Mic92) [inline flake.nix:61]: @Luis-Hebendanz any other sanitizer we should add?

Comment (Luis-Hebendanz) [reply flake.nix:61]: Ah yes I would add "-fsanitize=integer" too to catch unsigned integer overflow and integer implicit unsigned integer truncation :)

Comment (Mic92) [reply flake.nix:61]: Oh. Is this not covered yet by the undefined behavior sanitizer?

Comment (Luis-Hebendanz) [reply flake.nix:61]: No because unsigned integer overflow & truncation is not undefined behavior :)

**Pull request merged**